src/main/java/me/chrr/scribble/mixin/BookEditScreenMixin.java

```diff
@@ -497,31 +597,91 @@ private void setPageContent(String newContent) {
         Scribble.LOGGER.warn("setPageContent() was called, but ignored.");
     }
 
+    @Inject(method = "charTyped", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/util/SelectionManager;insert(Ljava/lang/String;)V"), cancellable = true)
+    private void charTypedEditMode(char chr, int modifiers, CallbackInfoReturnable<Boolean> cir) {
+        Command command = new InsertTextCommand<>(this, currentPageSelectionManager, Character.toString(chr));
+        commandManager.execute(command);
+        cir.setReturnValue(true);
+        cir.cancel();
+    }
+
     @Inject(method = "keyPressedEditMode", at = @At(value = "HEAD"), cancellable = true)
     private void keyPressedEditMode(int keyCode, int scanCode, int modifiers, CallbackInfoReturnable<Boolean> cir) {
-        // Copy/cut/paste without formatting when SHIFT is held down.
-        if (hasControlDown() && hasShiftDown() && !hasAltDown()) {
-            if (keyCode == GLFW.GLFW_KEY_C) {
-                this.getRichSelectionManager().copyWithoutFormatting();
-            } else if (keyCode == GLFW.GLFW_KEY_X) {
-                this.getRichSelectionManager().cutWithoutFormatting();
-            } else if (keyCode == GLFW.GLFW_KEY_V) {
-                this.getRichSelectionManager().pasteWithoutFormatting();
+        // Override copy-without-formatting shortcut
+        if (hasControlDown() && hasShiftDown() && !hasAltDown() && keyCode == GLFW.GLFW_KEY_C) {
+            getRichSelectionManager().copyWithoutFormatting();
+            cir.setReturnValue(true);
+            cir.cancel();
+            return;
+        }
+
+        // Override CUT and CUT-without-formatting shortcut
+        if (hasControlDown() && !hasAltDown() && keyCode == GLFW.GLFW_KEY_X) {
+            Command command = new ActionCommand<>(this, () -> {
+                if (hasShiftDown()) {
+                    getRichSelectionManager().cutWithoutFormatting();
+                } else {
+                    getRichSelectionManager().cut();
+                }
+            });
```

Comment:
This should just be a `DeleteTextCommand` I think, redoing a cut shouldn't alter your clipboard.

Reply:
Though that doesn't really work with how the command system works.. might want to add a way to add commands to the commandmanager without executing them somehow, and then performing the cut manually.